### ISSUE TITLE
Question about ensemble vcfs 

### ISSUE THREAD

**Issue by tb08** [chapmanb/bcbio-nextgen#2196]
Hello Brad et al.,
I am using bcbio v1.07a, ensemble tumor only variant calling. I noticed that the ensemble vcf in the final upload directory:

> final/2017-12-20_PROJECT/mysample-ensemble-annotated.vcf.gz

 is not the latest one from the pipeline:

> work/ensemble/mysample/mysample-ensemble-vepeffects.vcf.gz

but instead matches this one:

>  work/ensemble/mysample/mysample-ensemble.vcf.gz

which also contains vep annotations but the ones from before the decomposition of variants. I suspect this is not the intended behaviour?
Moreover, when I look in the mysample-ensemble-vepeffects.vcf there is a small formatting problem as there is a missing semicolon in the INFO field between CALLERS and the next INFO, for example:

> BaseQRankSum=-0.703;CALLERS=freebayes,vardict,varscanClippingRankSum=0;DB;

I think this might actually be a VEP issue, as I have just observed the same running VEP outside of bcbio.

Many thanks for all the great work.

**Comment by tb08** [comment]
I found another issue that arises from the new decomposition before ensemble recall. The decompose step now turns the soft filters in the individual caller vcfs into hard filters (the variants are dropped). This is because ensemble.py calls the normalize function with the option passonly=True. which breaks the "use_filtered: true" functionality.
I think this is where the problem arises, in ensemble.py, line 40:
` vrn_files = [normalize.normalize(f, data, passonly=True, rerun_effects=False) for f in vrn_files] `
I find the use_filtered:true option very useful because it allows applying different filters from the default ones in bcbio.

**Comment by chapmanb** [comment]
Thank you for all this feedback and for reporting the issues. I've been digging into this and having trouble replicating the initial behavior you describe. I only have tests setup for snpEff so have been using that instead of VEP, but appear to get the final re-annotated, post decomposition, input in the final directory. Would you be able to share your configuration so I can see if I can replicate directly? Do you see the same problematic behavior if you use snpEff instead of VEP?

For the semicolon issue, it sounds like that gets caused during VEP annotation, is that right? If so, @matthdsm just integrated a new VEP version so it might be worth updating and seeing if we can replicate outside of bcbio so we can report upstream to the VEP maintainers.

Finally, I pushed a fix for the normalization/use_filtered issue where we'll now respect that flag.

Thanks again for all this help debugging and sorry for the inability to spot a fix for the first two problems. I'll work more on setting up to reproduce and hopefully we can shake out the underlying issues.

**Comment by tb08** [comment]
Hi Brad,
Thanks for the reply and the fix for use_filtered.
About the first issue, what config info would you need? I unfortunately cannot share data but you will find below the template yaml I use for my bcbio run. Thing is, I have been running different versions of bcbio in this analysis, maybe a clean re-run of all the variant calling is the way to go? (individual variant caller analyses were run with 1.06 while ensemble was rerun with 1.07).

Yes I think the semicolon issue happens after VEP annotation. I think it only happens when VEP is run on a file that already has a CSQ field and this CSQ field is not the last one in the INFO section because the missing semicolon location is always exactly where the CSQ field used to be before reannotating. Probably something about the way VEP removes the existing CSQ before appending the new one at the end of the INFO section. Having fields after CSQ seems to happen only starting with -decomposed.vcf.gz VCFs. I cannot ascertain it right now as I am upgrading bcbio and it's still working on getting VEP to v 91 and the downloads are taking a while, I will try running ensemble again and see if these glitches are still happening and whether it's only with vep and I'll report back here. Thanks again for all the help.

```
# Tumor only variant calling
---
details:
  - analysis: variant2
    genome_build: GRCh37
    algorithm:
     aligner: bwa
     mark_duplicates: true
     recalibrate: false
     realign: false
     remove_lcr: true
     platform: illumina
     quality_format: standard
     min_allele_fraction: 2
     variantcaller: [gatk-haplotype, freebayes, vardict, varscan]
     ensemble:
      numpass: 1
      use_filtered: true
     tools_off: [gemini,tumoronly-prioritization,upload_alignment]
     effects: vep
     effects_transcripts: canonical
     tools_on: [qualimap]
     variant_regions: ./regions_GRCh37.bed
upload:
    dir: ../final
```

**Comment by tb08** [comment]
Hello again,
So I was able to confirm that the semicolon issue is related to vep and is still happening with the latest version (v91). I created a mock VCF with just two variants. One for which a CSQ field was present in the middle of some other INFO fields and one for which a CSQ was present as the last INFO section. When running VEP on this vcf, the first variant has a missing semicolon where the CSQ field used to be while for the second variant everything is fine, so the problem is on how VEP removes the existing CSQ field when in it's not in last position in the INFO. This does not happen for snpEff with the ANN field. 

For the first issue, I reran using snpEff for the annotation and the same problem happens: of the two VCFs present in the work/ensemble directory, the one that gets uploaded to the final directory is not the latest one (mysample-ensemble-effects.vcf.gz) but instead is the one which still has the annotations from the vcf of the first caller before decomposition (mysample-ensemble.vcf.gz).

Wishing everybody a Merry Christmas!

**Comment by chapmanb** [comment]
Thanks for all the additional details and helpful feedback. I pushed a fix that should avoid the CSQ issue by removing existing annotations when normalizing, allowing them to be added back in cleanly.

I'm still having trouble replicating your first problem and feel like I'm missing something. To double check, are you basing the missing annotations on the contents of the file, or the file name? The outputs in the final directory are often renamed to simplify them, since the `-effects` addition is more of an internal processing step and we don't need to expose it in the final file. Does that explain what you're seeing?

Thanks again for all this helpful discussion and feedback.

**Comment by tb08** [comment]
Hi Brad,
Thanks for looking into this. I saw the filename change but I am basing my observation on the content of the VCF.
The easiest way to see it is to look at the annotation commands recorded in the VCF headers:
`zcat final/2017-12-23/mySample-ensemble-annotated.vcf.gz | grep SnpEffCmd`
returns only one command (the one used to annotate the first caller's VCF, in my case gatk-haplotype).
while 
`zcat work/ensemble/mySample/mySample-ensemble-effects.vcf.gz | grep SnpEffCmd`
returns two commands, the first one is again the one for the annotation of the first caller's VCF while the second one is the latest one on the ensemble vcf file.
I just retried after updating bcbio to include your latest commit fixing the VEP semicolon glitch by removing the annotations from the decomposed files and this just confirms it: now the file that is uploaded in my final directory does not contain any annotations anymore while annotations are present in the work/ensemble/mysample-ensemble-effects.vcf.gz file.

**Comment by chapmanb** [comment]
Bruno -- thanks for following up and confirming it's not just a miscommunication issue. I've been unable to replicate with a bunch of different batching strategies matching your input configuration above, so am stuck right now trying to figure out what to fix. All of my tests have the annotated file and I'm missing the point in the code right now where it's going down the wrong path for you.

What type of batching strategy are you using for these samples? Are they run individually or in batches of multiple samples? I suspect something is going wrong with the population organization part of the code so maybe I am not properly replicating how you batch. Would you also be able to share the output of:
```
ls -lh gemini/*-ensemble*
```
which would help identify where the grouped ensemble calls point to. Thanks again for the patience debugging this.

**Comment by tb08** [comment]
Hi Brad,
I found the issue and it's not a bcbio bug. When I looked into the gemini directory as you asked I saw that symlinks for the ensemble VCFs pointed to mysample-ensemble.vcf.gz files, and that was because they had been created during my initial run of bcbio which did not have the extra decomposition yet and for which no reannotation was therefore performed.
Sorry for not catching this, I hadn't realized the gemini directory was still relevant as I had gemini deactivated with tools_off so I didn't realize it was necessary to delete it when I was re-running. I have now rerun on one sample after removing the gemini directory and everything worked as it should.
Sorry again for taking some of your time for this and thank you very much for all the patient help.
